src/System.IO.Compression/src/System/IO/Compression/DeflateStream.cs

```diff
@@ -0,0 +1,644 @@
+            EnsureNotDisposed();
+
+            int bytesRead;
+            int currentOffset = offset;
+            int remainingCount = count;
+
+            while (true)
+            {
+                bytesRead = inflater.Inflate(array, currentOffset, remainingCount);
+                currentOffset += bytesRead;
+                remainingCount -= bytesRead;
+
+                if (remainingCount == 0)
+                {
+                    break;
+                }
+
+                if (inflater.Finished())
+                {
+                    // if we finished decompressing, we can't have anything left in the outputwindow.
+                    Debug.Assert(inflater.AvailableOutput == 0, "We should have copied all stuff out!");
+                    break;
+                }
+
+                Debug.Assert(inflater.NeedsInput(), "We can only run into this case if we are short of input");
+
+                int bytes = _stream.Read(buffer, 0, buffer.Length);
+                if (bytes == 0)
+                {
+                    break;      //Do we want to throw an exception here?
+                }
+
+                inflater.SetInput(buffer, 0, bytes);
+            }
+
+            return count - remainingCount;
+        }
+
+        private void ValidateParameters(byte[] array, int offset, int count)
+        {
+            if (array == null)
+                throw new ArgumentNullException("array");
+
+            if (offset < 0)
+                throw new ArgumentOutOfRangeException("offset");
+
+            if (count < 0)
+                throw new ArgumentOutOfRangeException("count");
+
+            if (array.Length - offset < count)
+                throw new ArgumentException(SR.InvalidArgumentOffsetCount);
+        }
+
+        private void EnsureNotDisposed()
+        {
+            if (_stream == null)
+                throw new ObjectDisposedException(null, SR.ObjectDisposed_StreamClosed);
+        }
+
+        private void EnsureDecompressionMode()
+        {
+            if (_mode != CompressionMode.Decompress)
+                throw new InvalidOperationException(SR.CannotReadFromDeflateStream);
+        }
+
+        private void EnsureCompressionMode()
+        {
+            if (_mode != CompressionMode.Compress)
+                throw new InvalidOperationException(SR.CannotWriteToDeflateStream);
+        }
+
+        public override Task<int> ReadAsync(Byte[] array, int offset, int count, CancellationToken cancellationToken)
+        {
+            EnsureDecompressionMode();
+
+            // We use this checking order for compat to earlier versions:
+            if (asyncOperations != 0)
+                throw new InvalidOperationException(SR.InvalidBeginCall);
+
+            ValidateParameters(array, offset, count);
+            EnsureNotDisposed();
+
+            if (cancellationToken.IsCancellationRequested)
+            {
+                return TaskHelpers.FromCancellation<int>(cancellationToken);
```

Comment:
I realize you're just porting code, so this can be addressed separately, but TaskHelpers should no longer be needed here... this can just be Task.FromCancellation.